File: HW1_4/SpecialSequence.java
import java.util.*;

public class SpecialSequence {
   public static long Occur(int n) {
        return (long)n * (long)Math.sqrt(n) + (long)Math.ceil(n / 2.0);
    }
    
    public static int binarySearch(long[] arr, int l, int r, long x) {
        if (r >= l) {
            int mid = l + (r - l) / 2;
            //System.out.println(mid + " " + arr[mid] + " " + x);
    
            if (arr[mid] == x) return mid;
            else if (arr[mid] > x)
                return binarySearch(arr, l, mid - 1, x);
            else return binarySearch(arr, mid + 1, r, x);
        }

        return arr[l] >= x ? l : l - 1;
    }
    public static void main(String args[] ) throws Exception {
        Scanner sc = new Scanner(System.in);
        
        ArrayList<Long> arrlist = new ArrayList<Long>();

        long sum = 0; int i = 1;
        while(sum < Math.pow(10, 13)) {
            sum += Occur(i++);
            arrlist.add(sum);
        }

        long[] arr = new long[arrlist.size()];
        for(i = 0; i < arr.length; ++i) arr[i] = arrlist.get(i);


        int q = sc.nextInt();
        long l, r;
        while(q-- > 0) {
            l = sc.nextLong(); r = sc.nextLong();

            //int indexL = Collections.binarySearch(arrlist, l);
            //int indexR = Collections.binarySearch(arrlist, r);
            
            int indexL = binarySearch(arr, 0, arr.length, l);
            int indexR = binarySearch(arr, 0, arr.length, r);

            indexL = indexL < 0 ? -indexL - 1 : indexL;
            indexR = indexR < 0 ? -indexR - 1 : indexR;
            
            System.out.println(indexR - indexL + 1);
        }
    }
}

File: HW1_4/SpecialSequence.cpp
#include <iostream>
#include <cmath>
#define MAX 500000

long long int Occur(long long int i) {
	return i * (int)sqrt(i) + ceil(i / 2.0);
}

int binarySearch(long long int* arr, int l, int r, long long int x) {
    if (r >= l) {
        int mid = l + (r - l) / 2;
 
        if (arr[mid] == x)
            return mid;
 
        if (arr[mid] > x)
            return binarySearch(arr, l, mid - 1, x);
 
        return binarySearch(arr, mid + 1, r, x);
    }

    return arr[l] >= x ? l : l - 1;
}
 

int main() {
	long long int sum = 0;
	long long int* arr = new long long int[MAX];
	for(int i = 0; i < MAX; ++i) {
		sum += Occur(i);
		arr[i] = sum;
	}

	long long int q, l, r;

	std::cin >> q;

	while(q--) {
		std::cin >> l >> r;

		int rIdx = binarySearch(arr, 0, 500000, r);
		int lIdx = binarySearch(arr, 0, 500000, l);

		std::cout << rIdx - lIdx + 1 << std::endl;
	}

	return 0;
}
